### PR TITLE
bpf: Remove `__always_inline`

### PR DESCRIPTION
**Pull request by aditighag**
...This could simplify ifdefs and related `unused function` warnings from clang.

### CODE REVIEW

File: bpf/bpf_host.c [hunk 1]
```diff
@@ -58,12 +58,12 @@
 #define FROM_HOST_FLAG_NEED_HOSTFW (1 << 1)
 #define FROM_HOST_FLAG_HOST_ID (1 << 2)
 
-static __always_inline bool allow_vlan(__u32 __maybe_unused ifindex, __u32 __maybe_unused vlan_id) {
+static  bool allow_vlan(__u32 __maybe_unused ifindex, __u32 __maybe_unused vlan_id) {
```
Comment (lmb) [inline bpf/bpf_host.c:61]: Nit: there are a bunch of extra spaces here.

**Review by lmb — COMMENTED** [review]
The way I understand it, we need `always_inline` because otherwise the compiler might emit bpf2bpf calls instead of inlining. AFAIK we can't use those (yet) since combining bpf2bpf calls with tail calls is only supported as of recent 6.x kernels. Has that changed?

**Comment by joestringer** [comment]
@lmb thanks, that was the context we were looking for. The topic came up at LSF/MM/BPF that with a modern environment, we shouldn't need to always inline functions. I had suggested to just try this out to understand at which point we could start to get rid of these.

Poking around in https://github.com/cilium/cilium/pull/32533, I see a few other complications with a naive removal of these annotations ranging from functions with >5 args (currently OK because the arguments are never passed between functions due to inlining), unused function complaints (likely related to the macro configurations making a function unused in some configurations), some build-time constraints which may naively be made into runtime checks leading to failed compilation because we can't build a BPF program to call `abort()` to some issues around passing arguments from a static function to a non-static function. So even when we only support the relevant kernel versions, there'll be a few hoops for us to jump through to remove these.

**Comment by aditighag** [comment]
> The way I understand it, we need `always_inline` because otherwise the compiler might emit bpf2bpf calls instead of inlining. AFAIK we can't use those (yet) since combining bpf2bpf calls with tail calls is only supported as of recent 6.x kernels. Has that changed?

@lmb Yes, indeed! I reached the same conclusion after seeing the failures, but I was missing the context around the kernel version that supported both bpf2bpf calls and tail calls. Thanks for the context! 

Btw I was also looking at the cilium docs, specifically this note - 

```
Aside from BPF helper calls and BPF tail calls, a more recent feature that has been added to the BPF core infrastructure is BPF to BPF calls. Before this feature was introduced into the kernel, a typical BPF C program had to declare any reusable code that, for example, resides in headers as always_inline such that when LLVM compiles and generates the BPF object file all these functions were inlined and therefore duplicated many times in the resulting object file, artificially inflating its code size:

The main reason why this was necessary was due to lack of function call support in the BPF program loader as well as verifier, interpreter and JITs. Starting with Linux kernel 4.16 and LLVM 6.0 this restriction got lifted and BPF programs no longer need to use always_inline everywhere. 
```

Looks like the docs need an update.

**Comment by lmb** [comment]
I made a little table here: https://github.com/cilium/cilium/issues/30456 v6.0 + a RHEL bump is the minimum before we can attempt this again.

**Pull request closed without merging**